File: atcoder/educational_dp/4/b_frog2.py
import sys
from typing import List

import numpy as np

def solve (N: int, K: int, h: List[int]) -> int:

    dp = np.zeros(N, dtype=int)
    h = np.array(h)
    for i in range(1, N):
        dp[i] = min(dp[max(0, i-K): i] + np.abs(h[i] - h[max(0, i-K): i]))

    return dp[N-1]

def main () -> None:

    input = sys.stdin.readline

    N, K = [int(x) for x in input().split()]
    h = [int(x) for x in input().split()]

    result = solve(N, K, h)
    print(result)

    return

if __name__ == '__main__':
    main()

File: atcoder/educational_dp/3/b_frog2.cc
#include <iostream>
#include <vector>

using namespace std;
typedef long long ll;
const ll INF =  1LL << 60;

int N;
int K;
vector<int> h;

vector<ll> dp;

int solve (void) {
  dp[0] = 0;
  for (int i = 0; i < N; i++) {
    for (int j = 1; i + j < N && j <= K; j++) {
      dp[i + j] = min(dp[i + j], dp[i] + abs(h[i + j] - h[i]));
    }
  }

  return dp[N - 1];
}

int main (void) {
  cin >> N;
  cin >> K;
  h.assign(N, 0);
  dp.assign(N, INF);
  for (int i = 0; i < N; i++) {
    cin >> h[i];
  }
  int result = solve();
  cout << result << endl;
}
